File: android/lib/services/sync_orchestrator.dart
import 'dart:async';

import 'package:shared_preferences/shared_preferences.dart';

import '../models/obsyncian_config.dart';
import 'dynamodb_service.dart';
import 's3_sync_service.dart';

/// Current state of the sync engine.
enum SyncState { idle, syncing, error, offline }

/// Result of a sync cycle, with counts of files changed.
class SyncResult {
  final int uploaded;
  final int downloaded;
  final int deleted;

  bool get hadChanges => uploaded > 0 || downloaded > 0 || deleted > 0;

  const SyncResult({
    required this.uploaded,
    required this.downloaded,
    required this.deleted,
  });

  const SyncResult.noChanges()
      : uploaded = 0,
        downloaded = 0,
        deleted = 0;
}

/// Orchestrates sync operations between local files and S3/DynamoDB.
///
/// This mirrors the Go app's `handleSyncAsync` logic in `sync_logic.go`:
///   1. Get latest cloud sync from DynamoDB
///   2. If table empty -> sync up, create user
///   3. If user doesn't exist -> create user, sync down
///   4. If another user synced more recently -> sync down
///   5. Only if we did NOT sync down: dry-run local vs S3 -> sync up if changes
///   6. Update DynamoDB timestamp after sync up
///
/// The dry-run is skipped after a sync-down to prevent a race condition where
/// S3 changes between the sync-down and dry-run would be misinterpreted as
/// local changes, causing the mobile to overwrite newer cloud content.
class SyncOrchestrator {
  static const _lastSyncedTsKey = 'obsyncian_last_synced_timestamp';

  final S3SyncService _s3Service;
  final DynamoDBService _dynamoDBService;

  late ObsyncianConfig _config;

  SyncState _state = SyncState.idle;
  String _lastSyncedTimestamp = '';
  bool _isSyncQueued = false;

  /// Stream of log messages for the UI.
  final _logController = StreamController<String>.broadcast();
  Stream<String> get logStream => _logController.stream;

  /// Stream of state changes.
  final _stateController = StreamController<SyncState>.broadcast();
  Stream<SyncState> get stateStream => _stateController.stream;

  SyncState get state => _state;
  String get lastSyncedTimestamp => _lastSyncedTimestamp;

  SyncOrchestrator({
    required S3SyncService s3Service,
    required DynamoDBService dynamoDBService,
  })  : _s3Service = s3Service,
        _dynamoDBService = dynamoDBService;

  Future<void> configure(ObsyncianConfig config) async {
    _config = config;
    _s3Service.configure(config);
    _s3Service.onProgress = _log;
    _dynamoDBService.configure(config);
    await _loadLastSyncedTimestamp();
  }

  /// Run a full sync cycle. If a sync is already in progress, queues another
  /// cycle to run when the current one completes.
  /// Returns a [SyncResult] with counts of files changed, or null if the
  /// sync was queued (not executed immediately).
  Future<SyncResult?> sync() async {
    if (_state == SyncState.syncing) {
      _isSyncQueued = true;
      _log('Sync already in progress, queued another cycle.');
      return null;
    }

    _setState(SyncState.syncing);

    SyncResult result;
    try {
      result = await _handleSync();
      _setState(SyncState.idle);
    } catch (e) {
      _log('Sync error: $e');
      _setState(SyncState.error);
      rethrow;
    }

    // If another sync was requested while we were busy, run it now.
    if (_isSyncQueued) {
      _isSyncQueued = false;
      sync(); // fire-and-forget; it will set state itself
    }

    return result;
  }

  Future<SyncResult> _handleSync() async {
    _log('Last local sync: ${DateTime.now()}');
    _log('Config: region=${_config.awsRegion}, bucket=${_config.cloud}, '
        'key=${_config.credentials.key.substring(0, 4)}...');

    // 1. Get latest cloud sync
    _log('Scanning DynamoDB for latest sync...');
    final latestSync = await _dynamoDBService.getLatestSync();

    if (latestSync == null) {
      // Table is empty — this must be the first device ever.
      _log('Table is empty. Syncing up...');
      final diff = await _s3Service.syncUp();
      await _dynamoDBService.createUser(_config.id);
      await _dynamoDBService.updateTimestamp(_config.id);
      _log('Finished syncing up (new table).');
      return SyncResult(
        uploaded: diff.toUpload.length,
        downloaded: 0,
        deleted: diff.toDelete.length,
      );
    }

    _log('Latest cloud sync: ${latestSync.timestamp} by ${latestSync.userId}');

    // 2. Check if our user exists
    final ourItem = await _dynamoDBService.getUser(_config.id);

    if (ourItem == null) {
      _log('User ${_config.id} not found in table. Creating & syncing down...');
      await _dynamoDBService.createUser(_config.id);
      _log('Syncing down from S3...');
      final diff = await _s3Service.syncDown();
      await _saveLastSyncedTimestamp(latestSync.timestamp);
      _log('Finished syncing down (new user).');
      return SyncResult(
        uploaded: 0,
        downloaded: diff.toDownload.length,
        deleted: diff.toDelete.length,
      );
    }

    // 3. Check if we need to sync down
    final needsSyncDown = _config.id != latestSync.userId &&
        latestSync.timestamp.compareTo(ourItem.timestamp) >= 0 &&
        _lastSyncedTimestamp.compareTo(latestSync.timestamp) < 0;

    if (needsSyncDown) {
      _log('Not synced with Cloud. Syncing down from S3...');
      final diff = await _s3Service.syncDown();
      await _saveLastSyncedTimestamp(latestSync.timestamp);
      _log('Finished syncing down.');
      return SyncResult(
        uploaded: 0,
        downloaded: diff.toDownload.length,
        deleted: diff.toDelete.length,
      );
    }

    _log('Already synced with Cloud.');

    // 4. Check for local changes via dry-run
    _log('Checking for local changes (dry-run S3 diff)...');
    final diff = await _s3Service.dryRunSyncUp();

    if (diff.hasChanges) {
      _log('Local changes detected: $diff. Syncing up...');
      await _s3Service.syncUp();
      await _dynamoDBService.updateTimestamp(_config.id);
      _log('Finished syncing up.');
      return SyncResult(
        uploaded: diff.toUpload.length,
        downloaded: 0,
        deleted: diff.toDelete.length,
      );
    }

    _log('No local changes to sync.');
    return const SyncResult.noChanges();
  }

  void _setState(SyncState newState) {
    _state = newState;
    _stateController.add(newState);
  }

  void _log(String message) {
    _logController.add(message);
  }

  Future<void> _loadLastSyncedTimestamp() async {
    final prefs = await SharedPreferences.getInstance();
    _lastSyncedTimestamp = prefs.getString(_lastSyncedTsKey) ?? '';
  }

  Future<void> _saveLastSyncedTimestamp(String ts) async {
    _lastSyncedTimestamp = ts;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_lastSyncedTsKey, ts);
  }

  void dispose() {
    _logController.close();
    _stateController.close();
  }
}

File: android-rn/src/services/syncOrchestrator.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ObsyncianConfig } from '../models/config';
import { s3SyncService } from './s3SyncService';
import { dynamoDBService } from './dynamoDBService';
import { fileStatCache } from './fileStatCache';
import { ensureAndroidVaultStorageAccess } from './androidStoragePermission';

const LAST_SYNCED_TIMESTAMP_KEY = 'obsyncian_last_synced_timestamp';

export type SyncState = 'idle' | 'syncing' | 'error' | 'offline';

type OnStateChangeCallback = (state: SyncState) => void;
type OnLogCallback = (msg: string) => void;

class SyncOrchestrator {
  private isSyncing = false;
  private isSyncQueued = false;
  private lastSyncedTimestamp: string = '';
  private onStateChange: OnStateChangeCallback | null = null;
  private onLog: OnLogCallback | null = null;
  private isOnline = true;
  private vaultPath: string = '';

  async init(config: ObsyncianConfig, vaultPath: string, isOnline: boolean): Promise<void> {
    this.vaultPath = vaultPath;
    this.isOnline = isOnline;
    const onLog = (msg: string) => this.log(msg);

    // Load last synced timestamp
    const saved = await AsyncStorage.getItem(LAST_SYNCED_TIMESTAMP_KEY);
    if (saved) {
      this.lastSyncedTimestamp = saved;
      this.log(`[Sync] Loaded last synced timestamp: ${saved}`);
    }

    // Init services
    await s3SyncService.init(config, vaultPath, onLog);
    await dynamoDBService.init(config, onLog);
    await fileStatCache.load();
  }

  setOnStateChange(callback: OnStateChangeCallback): void {
    this.onStateChange = callback;
  }

  setOnLog(callback: OnLogCallback): void {
    this.onLog = callback;
  }

  setIsOnline(isOnline: boolean): void {
    this.isOnline = isOnline;
  }

  async sync(deviceId: string): Promise<void> {
    if (this.isSyncing) {
      this.log('[Sync] Already syncing, queueing next sync');
      this.isSyncQueued = true;
      return;
    }

    this.isSyncing = true;
    this.changeState('syncing');

    try {
      await this.handleSync(deviceId);

      if (this.isSyncQueued) {
        this.log('[Sync] Queued sync triggered');
        this.isSyncQueued = false;
        await this.sync(deviceId);
      }

      this.changeState('idle');
    } catch (error) {
      this.log(`[Sync] Error: ${error}`);
      this.changeState('error');
    } finally {
      this.isSyncing = false;
    }
  }

  private async handleSync(deviceId: string): Promise<void> {
    this.log('[Sync] Starting sync cycle');

    if (!this.isOnline) {
      this.log('[Sync] Offline, skipping sync');
      this.changeState('offline');
      return;
    }

    await ensureAndroidVaultStorageAccess(this.vaultPath, (msg) => this.log(msg));

    // 1. Scan DynamoDB for latest timestamp
    const latestSync = await dynamoDBService.getLatestSync();

    // 2. If table is empty
    if (!latestSync) {
      this.log('[Sync] Table empty, syncing up and creating user');
      await s3SyncService.syncUp();
      await dynamoDBService.createUser(deviceId);
      await dynamoDBService.updateTimestamp(deviceId);
      return;
    }

    // 3. Check if our user exists
    const ourUser = await dynamoDBService.getUser(deviceId);

    if (!ourUser) {
      this.log('[Sync] User not found in table, creating and syncing down');
      await dynamoDBService.createUser(deviceId);
      await s3SyncService.syncDown();
      // Remember the DynamoDB cloud revision we pulled (same as Flutter).
      await this.saveLastPulledCloudRevision(latestSync.timestamp);
      return;
    }

    // 4. Sync down only if cloud has a newer revision than the last one we pulled (YYYYMMDD… UTC strings).
    const shouldSyncDown =
      latestSync.userId !== deviceId &&
      latestSync.timestamp >= ourUser.timestamp &&
      (!this.lastSyncedTimestamp || this.lastSyncedTimestamp < latestSync.timestamp);

    if (shouldSyncDown) {
      this.log(
        `[Sync] Cloud has a newer revision (${latestSync.userId} @ ${latestSync.timestamp}), syncing down`,
      );
      await s3SyncService.syncDown();
      await this.saveLastPulledCloudRevision(latestSync.timestamp);
      return;
    }

    // 5. Dry run to check for local changes
    this.log('[Sync] Checking for local changes');
    const diff = await s3SyncService.dryRun();

    if (diff.toUpload.length > 0 || diff.toDelete.length > 0) {
      this.log('[Sync] Local changes detected, syncing up');
      await s3SyncService.syncUp();
      await dynamoDBService.updateTimestamp(deviceId);
    } else {
      this.log('[Sync] No local changes detected');
    }

    // Save file stat cache
    await fileStatCache.save();
  }

  /**
   * Persist the DynamoDB “latest cloud sync” timestamp we have fully applied locally.
   * Must be the same YYYYMMDDHHmmss string as in the Obsyncian table (not wall-clock).
   */
  private async saveLastPulledCloudRevision(timestamp: string): Promise<void> {
    this.lastSyncedTimestamp = timestamp;
    await AsyncStorage.setItem(LAST_SYNCED_TIMESTAMP_KEY, timestamp);
    this.log(`[Sync] Saved last pulled cloud revision: ${timestamp}`);
  }

  private log(msg: string): void {
    const timestamp = new Date().toLocaleTimeString();
    const logMsg = `[${timestamp}] ${msg}`;
    console.log(logMsg);
    if (this.onLog) {
      this.onLog(logMsg);
    }
  }

  private changeState(state: SyncState): void {
    if (this.onStateChange) {
      this.onStateChange(state);
    }
  }
}

export const syncOrchestrator = new SyncOrchestrator();
